File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.py
import sys
from collections import deque

input = sys.stdin.readline
N,M = map(int,input().split())
graph = [list(map(int,input().split())) for _ in range(N)]
visited = [[-1] * M for _ in range(N)]
dx = [-1,1,0,0]
dy = [0,0,-1,1]

def bfs(x,y):
    queue = deque()
    queue.append((x,y))
    visited[x][y] = 0
    while queue:
        x,y = queue.popleft()
        for i in range(4):
            nx = dx[i] + x
            ny = dy[i] + y
            if 0 <= nx < N and 0 <= ny < M and visited[nx][ny] == -1:
                if graph[nx][ny] == 0:
                    visited[nx][ny] = 0
                elif graph[nx][ny] == 1:
                    visited[nx][ny] = visited[x][y] +1
                    queue.append((nx,ny))
for i in range(N):
    for j in range(M):
        if graph[i][j] == 2 and visited[i][j] == -1:
            bfs(i,j)

for i in range(N):
    for j in range(M):
        if graph[i][j] == 0:
            print(0,end=' ')
        else:
            print(visited[i][j],end=' ')
    print()

File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.java
import java.io.*;
import java.util.*;


public class Main{

    static int[][] graph;
    static boolean[][] visited;
    static int[] dx = {-1,1,0,0};
    static int[] dy = {0,0,-1,1};

    static int N,M;

    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        graph = new int[N][M];
        visited = new boolean[N][M];
        for(int i=0; i<N; i++){
            st = new StringTokenizer(br.readLine());
            for(int j=0; j<M; j++){
                graph[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        int start = 0;
        int end = 0;
        for(int i=0; i<N; i++){
            for(int j=0; j<M; j++){
                if(graph[i][j] == 2){
                    start = i;
                    end = j;
                }
            }
        }
        graph[start][end] = 0;
        bfs(start,end);
        for(int i=0; i<N; i++){
            for(int j=0; j<M;j++){
                if(graph[i][j] !=0 && !visited[i][j]){
                    graph[i][j] = -1;
                }
            }
        }
        for(int i=0; i<N; i++){
            for(int j=0; j<M; j++){
                System.out.print(graph[i][j]+" ");
            }
            System.out.println();
        }
    }

    static void bfs(int x, int y){
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{x,y});
        visited[x][y] = true;
        while(!queue.isEmpty()){
            int[] poll = queue.poll();
            int CurX = poll[0];
            int CurY = poll[1];
            for(int i=0; i<4; i++){
                int nx = CurX + dx[i];
                int ny = CurY + dy[i];
                if(nx < 0 || ny < 0 || nx >= N || ny >= M){
                    continue;
                }
                if(!visited[nx][ny] && graph[nx][ny] != 0){
                    graph[nx][ny] = graph[CurX][CurY] + 1;
                    visited[nx][ny] = true;
                    queue.add(new int[]{nx,ny});
                }
            }
        }
    }
}